File: client-sdk/src/services/bulk.ts
/**
 * Bulk Scoring Service
 * Wraps /bulk endpoints for batch transaction scoring
 */

import { AxiosInstance } from 'axios';
import { EventEmitter } from '../events';
import { BaseService } from './base';

export type RiskLevel = 'low' | 'medium' | 'high' | 'critical';

export interface BulkScoringRequest {
  transactions: BulkTransaction[];
  options?: {
    includeLabels?: boolean;
    includeFactors?: boolean;
    parallelism?: number;
    timeout?: number;
  };
}

export interface BulkTransaction {
  id: string;
  from: string;
  to: string;
  amount: string;
  tokenAddress?: string;
  chainId?: number;
  metadata?: Record<string, any>;
}

export interface BulkScoringResult {
  jobId: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  results: TransactionScoreResult[];
  processedCount: number;
  failedCount: number;
  failures: { id: string; error: string }[];
  startedAt: string;
  completedAt?: string;
  processingTimeMs?: number;
}

export interface TransactionScoreResult {
  id: string;
  from: string;
  to: string;
  riskScore: number;
  riskLevel: RiskLevel;
  fromRisk: { score: number; level: RiskLevel };
  toRisk: { score: number; level: RiskLevel };
  labels?: { address: string; labels: string[] }[];
  factors?: { name: string; weight: number; value: number }[];
  allowed: boolean;
  reason?: string;
}

export interface BulkJobStatus {
  jobId: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  progress: number;
  processedCount: number;
  totalCount: number;
  estimatedTimeRemaining?: number;
  startedAt: string;
  updatedAt: string;
}

export class BulkService extends BaseService {
  constructor(http: AxiosInstance, events: EventEmitter) {
    super(http, events);
  }

  /**
   * Submit bulk scoring request
   */
  async submit(request: BulkScoringRequest): Promise<{
    jobId: string;
    status: string;
    estimatedCompletionTime: string;
  }> {
    const response = await this.http.post<{
      jobId: string;
      status: string;
      estimatedCompletionTime: string;
    }>('/bulk/submit', request);

    return response.data;
  }

  /**
   * Submit and wait for results (synchronous for small batches)
   */
  async score(request: BulkScoringRequest): Promise<BulkScoringResult> {
    const response = await this.http.post<BulkScoringResult>('/bulk/score', request);
    return response.data;
  }

  /**
   * Get job status
   */
  async getStatus(jobId: string): Promise<BulkJobStatus> {
    const response = await this.http.get<BulkJobStatus>(`/bulk/status/${jobId}`);
    return response.data;
  }

  /**
   * Get job results
   */
  async getResults(jobId: string, options?: {
    limit?: number;
    offset?: number;
    status?: 'success' | 'failed';
  }): Promise<{
    results: TransactionScoreResult[];
    total: number;
    hasMore: boolean;
  }> {
    const params = new URLSearchParams();
    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.offset) params.append('offset', options.offset.toString());
    if (options?.status) params.append('status', options.status);

    const response = await this.http.get<{
      results: TransactionScoreResult[];
      total: number;
      hasMore: boolean;
    }>(`/bulk/results/${jobId}?${params.toString()}`);

    return response.data;
  }

  /**
   * Cancel a bulk job
   */
  async cancel(jobId: string): Promise<{ cancelled: boolean; processedBeforeCancel: number }> {
    const response = await this.http.post(`/bulk/cancel/${jobId}`);
    return response.data;
  }

  /**
   * Get job history
   */
  async getHistory(options?: {
    limit?: number;
    offset?: number;
    status?: string;
    startDate?: Date;
    endDate?: Date;
  }): Promise<{
    jobs: BulkJobStatus[];
    total: number;
  }> {
    const params = new URLSearchParams();
    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.offset) params.append('offset', options.offset.toString());
    if (options?.status) params.append('status', options.status);
    if (options?.startDate) params.append('startDate', options.startDate.toISOString());
    if (options?.endDate) params.append('endDate', options.endDate.toISOString());

    const response = await this.http.get<{
      jobs: BulkJobStatus[];
      total: number;
    }>(`/bulk/history?${params.toString()}`);

    return response.data;
  }

  /**
   * Get bulk scoring statistics
   */
  async getStatistics(): Promise<{
    totalJobs: number;
    totalTransactionsProcessed: number;
    averageProcessingTime: number;
    successRate: number;
    byStatus: Record<string, number>;
  }> {
    const response = await this.http.get('/bulk/statistics');
    return response.data;
  }

  /**
   * Download results as CSV
   */
  async downloadResults(jobId: string): Promise<Blob> {
    const response = await this.http.get(`/bulk/download/${jobId}`, {
      responseType: 'blob'
    });
    return response.data;
  }

  /**
   * Retry failed transactions in a job
   */
  async retryFailed(jobId: string): Promise<{
    newJobId: string;
    transactionsToRetry: number;
  }> {
    const response = await this.http.post(`/bulk/retry/${jobId}`);
    return response.data;
  }
}

File: client-sdk-python/amttp/services/bulk.py
"""Bulk Scoring Service — wraps ``/bulk`` endpoints."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel

from amttp.services.base import BaseService


class BulkTransaction(BaseModel):
    from_address: str = ""
    to_address: str = ""
    amount: str = ""
    chain_id: int = 1


class TransactionScoreResult(BaseModel):
    from_address: str = ""
    to_address: str = ""
    risk_score: float = 0
    risk_level: str = "low"
    action: str = "allow"
    error: Optional[str] = None


class BulkScoringRequest(BaseModel):
    transactions: List[BulkTransaction] = []
    include_factors: bool = False


class BulkScoringResult(BaseModel):
    results: List[TransactionScoreResult] = []
    processed_count: int = 0
    failed_count: int = 0


class BulkJobStatus(BaseModel):
    job_id: str = ""
    status: str = "pending"
    progress: float = 0
    total: int = 0
    completed: int = 0
    failed: int = 0
    created_at: str = ""
    completed_at: Optional[str] = None


# ── service ───────────────────────────────────────────────────────────────────


class BulkService(BaseService):
    """Bulk scoring operations."""

    async def score(self, request: BulkScoringRequest) -> BulkScoringResult:
        """Score transactions in bulk (synchronous)."""
        resp = await self._http.post("/bulk/score", json=request.model_dump())
        resp.raise_for_status()
        return BulkScoringResult.model_validate(resp.json())

    async def submit_job(self, request: BulkScoringRequest) -> BulkJobStatus:
        """Submit an async bulk scoring job."""
        resp = await self._http.post("/bulk/submit", json=request.model_dump())
        resp.raise_for_status()
        return BulkJobStatus.model_validate(resp.json())

    async def get_job_status(self, job_id: str) -> BulkJobStatus:
        """Get bulk job status."""
        resp = await self._http.get(f"/bulk/job/{job_id}")
        resp.raise_for_status()
        return BulkJobStatus.model_validate(resp.json())

    async def get_job_results(self, job_id: str) -> BulkScoringResult:
        """Get bulk job results."""
        resp = await self._http.get(f"/bulk/job/{job_id}/results")
        resp.raise_for_status()
        return BulkScoringResult.model_validate(resp.json())

    async def cancel_job(self, job_id: str) -> Dict[str, Any]:
        """Cancel a running bulk job."""
        resp = await self._http.post(f"/bulk/job/{job_id}/cancel")
        resp.raise_for_status()
        return resp.json()
